cdisc_rules_engine/services/data_services/usdm_data_service.py

```diff
@@ -0,0 +1,310 @@
+        """
+        metadata: dict = self.read_metadata(dataset_name)
+        contents_metadata: dict = metadata["contents_metadata"]
+        metadata_to_return: VariableMetadataContainer = VariableMetadataContainer(
+            contents_metadata
+        )
+        return pandas.DataFrame.from_dict(metadata_to_return.to_representation())
+
+    @cached_dataset(DatasetTypes.CONTENTS.value)
+    def get_define_xml_contents(self, dataset_name: str) -> bytes:
+        """
+        Reads local define xml file as bytes
+        """
+        raise NotImplementedError(
+            "Can't use 'get_define_xml_contents' in USDMDataService!"
+        )
+
+    def read_metadata(self, dataset_name: str) -> dict:
+        np_json_type_map: dict = {"O": "string", "float64": "float"}
+        file_size = os.path.getsize(self.dataset_path)
+        file_name = extract_file_name_from_path_string(self.dataset_path)
+        file_metadata = {
+            "path": self.dataset_path,
+            "name": file_name,
+            "size": file_size,
+        }
+        dataset = self.__get_dataset(dataset_name)
+        measurer = vectorize(len)
+        contents_metadata = {
+            "variable_labels": dataset.columns.values.tolist(),
+            "variable_names": dataset.columns.values.tolist(),
+            "variable_formats": empty(dataset.shape[1], dtype=str).tolist(),
+            "variable_name_to_label_map": dict(zip(dataset, dataset)),
+            "variable_name_to_data_type_map": dict(
+                zip(dataset, dataset.dtypes.map(np_json_type_map))
+            ),
+            "variable_name_to_size_map": dict(
+                zip(dataset, measurer(dataset.values.astype(str)).max(axis=0))
+            ),
+            "number_of_variables": dataset.shape[1],
+            "dataset_label": dataset_name,
+            "dataset_length": dataset.shape[0],
+            "domain_name": dataset_name,
+            "dataset_name": dataset_name,
+            "dataset_modification_date": datetime.fromtimestamp(
+                os.path.getmtime(self.dataset_path)
+            ).isoformat(),
+        }
+
+        return {
+            "file_metadata": file_metadata,
+            "contents_metadata": contents_metadata,
+        }
+
+    def read_data(self, file_path: str) -> IOBase:
+        return open(file_path, "rb")
+
+    def get_datasets(self) -> List[dict]:
+        json = self._reader_factory.get_service("USDM").from_file(self.dataset_path)
+        return self.__get_datasets(json)
+
+    def __get_record_data(self, node: dict, parent="") -> dict:
+        if type(node) is dict:
+            flattened = {}
+            for key, value in node.items():
+                if type(value) is dict:
+                    flattened |= self.__get_record_data(value, f"{parent}{key}.")
+                elif type(value) is not list:
+                    flattened[f"{parent}{key}"] = value
```

Comment:
I was aware of this. I did not see it as an issue because when the dataframe is built, it includes the columns from both records.
On second look, I see one issue being that authors might be tempted to test:
```yaml
name: studyIdentifierScope.organizationLegalAddress
operator: empty
```
Both records will return true, even though it does exist for the non-null record. Either authors will need to specify non-object properties or look at the `null` dataset.
We could do something better to distinguish between these

Reply:
I've updated this so that populated dict and list properties will have a `True` value instead of `Null`.

Reply:
I've confirmed that this looks to be working correctly.  The only thing that now looks a bit strange is that you get `True` for an empty list (`[]`), which seems to imply that you should be able to find the list contents somewhere.

Reply:
This is what I would expect so that there is a way to represent empty lists. Do you have an example rule where this would cause a problem that can't be solved?

Reply:
The one I was looking at was `study.studyPhase.standardCodeAliases` - I saw `True`, interpreted it as "there's a definition for this" and was therefore expecting to find records in the `Code` dataset with `parent_rel = "standardCodeAliases"`

Reply:
do you think it is okay?